docs/extensions/dnsrecord.md

```diff
@@ -0,0 +1,123 @@
+# Contract: `DNSRecord` resources
+
+Every shoot cluster requires external DNS records that are publicly resolvable.
+The management of these DNS records requires provider-specific knowledge which is to be developed outside the Gardener's core repository.
+
+Currently, Gardener uses [`DNSProvider` and `DNSEntry` resources](dns.md). However, this introduces undesired coupling of Gardener to a controller that does not adhere to the Gardener extension contracts. Because of this, we plan to stop using `DNSProvider` and `DNSEntry` resources for Gardener DNS records in the future and use the `DNSRecord` resources described here instead. 
+
+## What does Gardener create DNS records for?
+
+### Internal domain name
+
+Every shoot cluster's kube-apiserver running in the seed is exposed via a load balancer that has a public endpoint (IP or hostname).
+This endpoint is used by end-users and also by system components (that are running in another network, e.g., the kubelet or kube-proxy) to talk to the cluster.
+In order to be robust against changes of this endpoint (e.g., caused due to re-creation of the load balancer or move of the DNS record to another seed cluster) Gardener creates a so-called *internal domain name* for every shoot cluster.
```

Comment:
It is not for every shoot cluster. If the seed does not support DNS, it will use the public endpoint of the Load Balancer. 
See https://gardener.cloud/documentation/references/core/#core.gardener.cloud/v1beta1.SeedSettingShootDNS.

Reply:
This was part of #4231